20140629_SigmaSGD.README

Motivation for this version:
It came to my attention that the old sigma genome and annotations that we were using (obtained from Robin Dowell in 2009 before the publication of the sigma genome)
have quite a few genome->annotation errors, including an off-by-one error that affects most of chromosome 8, but isn't so consistent that it could have been easily corrected.
A version of the Sigma1278 draft that was posted on SGD has almost none of these errors. The difference is now that reads are mapped to contigs rather than chromosome sequences,
but since the gene annotations are given with respect to the contigs this doesn't practically make a difference.

History:
- Made 20140629_SigmaSGD/ folder in the pipeline/genome_features/
- Made subdirectory fromSGD/ for the original files, used wget to download from SGD on 140629. Also downloaded chrmt.fsa and saccharomyces_cerevisiae.gff to use to build the mt annotations
- Made scripts directory SigmaSGDscripts/ to store scripts used for conversion of SGD files to each file type used by the pipeline (as necessary)
- Make output directory to contain all the files used by the pipeline Sigma_pipeline_files/

File creation:
1) Make a combined sigma genome+ mt genome fasta file:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/addMtFasta.py fromSGD/Sigma1278b_MIT_2009_ACVY01000000.fsa fromSGD/chrmt.fsa Sigma_pipeline_files/Sigma1278b_full.fsa

2) Make a combined sigma genome+ mt genome gff file:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/addMtGff.py fromSGD/Sigma1278b_ACVY01000000.gff fromSGD/saccharomyces_cerevisiae.gff Sigma_pipeline_files/Sigma1278b_full.gff

3) pickle the full sigma genome:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/pickleContigs.py Sigma_pipeline_files/Sigma1278b_full.fsa Sigma_pipeline_files/Sigma1278b_full.p

4) Make the .tab file (this file contains everything annoated to CDS, whether Characterized, Uncharacterized, or Dubious). The decisions/filtering step can be done later after the pipeline is run
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/convertGFF2tab.py Sigma_pipeline_files/Sigma1278b_full.gff Sigma_pipeline_files/Sigma1278b_full.tab

5) Make the splice junctions file:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/getSigmaSpliceJunctions.py Sigma_pipeline_files/Sigma1278b_full.tab Sigma_pipeline_files/Sigma1278b_full.juncs

6) Make the coding pickle file:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/makeCodingPickle.py Sigma_pipeline_files/Sigma1278b_full.gff Sigma_pipeline_files/Sigma1278b_full.p Sigma_pipeline_files/Sigma1278b_full_coding.p
correctATG 6438
incorrectATG 9
correctStop 6443
incorrectStop 4
div3 6441

7) Remove ORFs that overlap with the rRNA:
Made a list of all the ORFs that overlap with the rRNA, will remove these ORFs from the .tab and coding.p file
List of ORFs= rRNA_CDS_overlaps.txt
file ending= *_clean*
- Use the clean files in the pipeline, decide later for other analyses whether we want to filter out the non-verified ORFs
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/addremoveGenes.py Sigma_pipeline_files/Sigma1278b_full.tab Sigma_pipeline_files/Sigma1278b_full_coding.p Sigma_pipeline_files/Sigma1278b_full_clean -geneset Sigma_pipeline_files/rRNA_CDS_overlap.txt -remove
removed 6
Since there were 10 shown to overlap in SGD, I guess not all of them made it to the Sigma.gff file

8) Remove non-verified ORFs:
Make a list Verified ORFs:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/makeORFtypelist.py Sigma_pipeline_files/Sigma1278b_full.gff Sigma_pipeline_files/Sigma1278b_full_verifiedlist.txt -orf_class Verified

Make the _verified versions of these files:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/addremoveGenes.py Sigma_pipeline_files/Sigma1278b_full.tab Sigma_pipeline_files/Sigma1278b_full_coding.p Sigma_pipeline_files/Sigma1278b_full_verified -geneset Sigma_pipeline_files/Sigma1278b_full_verifiedlist.txt -keep

9) Make a list of Uncharacterized ORFs
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/makeORFtypelist.py Sigma_pipeline_files/Sigma1278b_full.gff Sigma_pipeline_files/Sigma1278b_full_uncharacterizedlist.txt -orf_class Uncharacterized
Combine the verified and uncharacterized ORF lists:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ cat Sigma_pipeline_files/Sigma1278b_full_verifiedlist.txt Sigma_pipeline_files/Sigma1278b_full_uncharacterizedlist.txt > Sigma_pipeline_files/Sigma1278b_full_Verifed_Uncharacterizedlist.txt
Make annotations corresponding to the verified and Uncharacterized ORFs:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD$ python SigmaSGDscripts/addremoveGenes.py Sigma_pipeline_files/Sigma1278b_full.tab Sigma_pipeline_files/Sigma1278b_full_coding.p Sigma_pipeline_files/Sigma1278b_full_verified_unchar -geneset Sigma_pipeline_files/Sigma1278b_full_Verifed_Uncharacterizedlist.txt -keep

10) Build corresponding bowtie indices:
mk@gilbertlab-prt1:~/pipeline_Seq/genome_features/20140629_SigmaSGD/Sigma_pipeline_files$ cp Sigma1278b_full.fsa ../../../scripts/bowtieindexes/20140629_sigmaSGD/
mk@gilbertlab-prt1:~/pipeline_Seq/scripts/bowtieindexes/20140629_sigmaSGD$ bowtie2-build Sigma1278b_full.fsa Sigma1278b_full

Build the rRNA index (use the RDN37-1 and RDN5-1 from SGD)
mk@gilbertlab-prt1:~/pipeline_Seq/scripts/bowtieindexes/20140629_sigmaSGD$ ~/TeamGilbert/programs/bowtie-0.12.5/bowtie-build rRNA_SGD.fa rRNA_SGD


